Biz/Model/UserLoginBaseForm.pm

```diff
@@ -113,6 +125,48 @@ sub validate {
     return;
 }
 
+sub validate_and_execute_ok {
+    my($self) = @_;
+    my($res) = shift->SUPER::validate_and_execute_ok(@_);
+    # Won't have a realm owner if the login field was invalid
+    return $res
+        unless my $owner = $self->unsafe_get('realm_owner');
+    my($errors) = $self->get_errors || {};
+    if ($errors->{'login'} && $errors->{'login'}->eq_locked_out) {
+        b_warn('locked out owner=', $owner);
+        # Return non-specific errror so potential attacker doesn't glean additional information
+        return {
+            method => 'server_redirect',
+            task_id => 'DEFAULT_ERROR_REDIRECT',
+            query => undef,
+        };
+    }
+    # Have to do this after form processing as the transaction gets rolled back upon form errors
```

Comment:
The comment might be better this way: We do not fall through, because form_errors will rollback, clearing the LoginAttempt. The server_redirect ensures the lockout message is shown to the user.

Reply:
I added this comment to point out that the login attempt can't be recorded at some earlier point, like in `validate`. Saying that we don't fall through isn't helpful to me. I understand (now) that `validate` wouldn't be the right place to record or modify data in general, but I didn't realize that before I started this PR.

Reply:
OK. However you want to document is fine. I would say validate is fine enough. It's more complicated to do it any other way, because you have to catch all cases.

Reply:
> I would say validate is fine enough.

Do you mean recording the login attempt is fine enough in `validate`? As I say, if there are form errors, then that record would get rolled back.

Reply:
The record shouldn't get rolled back when the password is incorrect. I think we might be able to avoid this by setting [stay_on_page](https://github.com/biviosoftware/perl-Bivio/blob/71d45f48cb1621d8d316c1bf04ab63b88b665821/Biz/FormModel.pm#L1033) when there are form_errors. I don't know if we use it this way in other places, but I think it makes sense to do it this way.

Reply:
Thanks.